src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs

```diff
@@ -35,6 +35,20 @@ public static class UnmanagedCallersOnlyDll
         public static extern int PInvokeMarkedWithUnmanagedCallersOnly(int n);
     }
 
+    private const string InvalidCSharpAssemblyName = "InvalidCSharp";
+
+    public static Type GetCallbacksType()
+    {
+        var asm = Assembly.Load(InvalidCSharpAssemblyName);
+        return asm.GetType("InvalidCSharp.Callbacks");
```

Comment:
The ProjectReferences seems to be working fine for test projects. For example: https://github.com/dotnet/runtime/blob/master/src/tests/baseservices/callconvs/TestCallingConventions.csproj#L11

Project reference would be even better than `Assembly.Load`.

Reply:
Yes, `ProjectReference`s do work for build but not to refer to the assembly. The typical issue is the assembly that contains `System.Object` isn't the same when compiled with `ilasm` as the one for `csc` and during assembly reference an error occurs. The only way I have found to circumvent this is to load during runtime.

The reason the referenced test works is because of the following:

https://github.com/dotnet/runtime/blob/c86733753da5ec1b2e670cad229f38a5bc113f6e/src/tests/baseservices/callconvs/CallFunctionPointers.il#L4-L8

This is fine for now, but when we version to .NET 6, this will break. The current approach is the only way I have found that continue to work through time.

Reply:
I do not think this should be a problem. What is the csc error that you are seeing? I know that csc was picky about situations like these in the past, but these checks were removed a while ago. They did not make sense for .NET Core nor for .NET Framework with binding redirects. 

It is fairly normal to have multiple different System.Runtime references in .NET Core world. For example, it is fully supported for project targeting net6 (that has System.Runtime 6.0) to reference project targeting net6 (that has System.Runtime 5.0.).

If you do not like hardcoding versions, you can just omit them. For example: https://github.com/jkotas/runtime/blob/5095f1f75f86c2aedaedf9e35457cb6635a3c989/src/tests/Regressions/coreclr/16355/boring.il#L4

Reply:
I was hitting the error about `System.Object` from assembly is not known. However, when I try it now I don't see that... this is unsettling because I did try that first. I'm not a fan of not understanding why I hit that error for 20 minutes changed to the runtime load but now it doesn't repro.

Reply:
Yep, that works just fine throughout. Any idea how I could have made that error occur? That is frustrating beyond belief. I will push up a new PR with the IL updates as well.

Reply:
> error about System.Object from assembly is not known.

This error is typically caused by missing System.Runtime reference, not by having incompatible System.Runtime references.

Reply:
Here is what I am seeing now.

Failure during build of C# test application:
```
error CS0012: The type 'Object' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Runtime, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
```

Call site in C#:
```CSharp
InvalidCSharp.Test.Invoke(null, n);
```

Snippet of IL code being called in C#:
```
.assembly extern System.Runtime { }
.assembly extern System.Runtime.InteropServices { }

.assembly InvalidCSharp { }

.class public auto ansi beforefieldinit InvalidCSharp.Test
    extends System.Object
{
    .method public hidebysig static 
        void Invoke (
            void* fptr,
            int32 n
        ) cil managed 
    {
        .maxstack 2
        .locals init ([0] native int ptr)

        nop
        ldarg.0
        stloc.0

        ldloc.0
        ldarg.1
        calli int32 *(int32)
        
        ret
    }
}
```

Reply:
Maybe C# compiler needs the public key tokens to match? You can try adding ` .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )`.

Reply:
Does the C# compiler command line have `/r:....\System.Runtime.dll` ?

Reply:
>Maybe C# compiler needs the public key tokens to match? You can try adding .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A ).

That did it. This is a very confusing issue. But it works now so yay.

>Does the C# compiler command line have /r:....\System.Runtime.dll ?

It does.

```
 /reference:D:\runtime\artifacts\bin\ref\net5.0\System.Runtime.dll
```